docs/channels/library_auth_reference/server-library-reference-specification.md

```diff
@@ -0,0 +1,234 @@
+// socketId: The socket identifier of the connected client.
+// userDataJSON: user data JSON for a presence channel subscription attempt.
+// Async callback: A list of channel summaries, or an error.
+func authenticate(channelName: String,
+                  socketId: String,
+                  userDataJSON: [String: Any]?,
+                  callback: Result<AuthenticationToken, PusherError>)
+```
+
+- The method parameters SHOULD be encapsulated into public containing types for typed languages where this is more in keeping with best practices for the development language.
+  - E.g. a `Channel` object that has  channel name `String` and channel type `enum` properties.
+- The library SHOULD specify the required public types to encapsulate the objects returned by these top-level API methods.
+  - If appropriate for the development language, lower-level primitives such as Strings, Arrays or Dictionaries SHOULD be used to encapsulate the returned data instead.
+
+## Configuration options
+
+- The library MUST provide a method to configure the connection to the Channels HTTP API.
+  - This could be a type encapsulating the configuration options, or an API method to provide them.
+- The library MUST accept the following required configuration parameters:
+  - An `app_id` string
+  - A `key` string
+  - A `secret` string
+- The library MUST accept the following optional configuration parameters:
+  - A `cluster` string
+  - A `host` string
+  - A `httpProxy` string
+  - A `port` string
+  - A `useTLS` boolean flag
+  - An `encryptionMasterKey` string
+
+## Documentation
+
+- The library MUST provide a comprehensive README
+  - The README MUST contain well-formatted code examples covering the public API of the library.
+  - The README MUST provide instructions for installation and configuration, as well as a link to full API documentation.
+  - The README MUST be kept up to date based on any changes to the public API.
+- The library MUST have its public API fully documented via code comments on the public interfaces.
+- The library SHOULD host comprehensive API documentation, automatically generated from the documentation comments discussed in the previous bullet point.
+
+## Error reporting
+
+- The library MUST report errors using an accepted “best practice” for the development language or platform.
+  - Examples of a suitable error reporting mechanism could be throwing catchable exceptions, returning errors using a delegate pattern, or logging errors to a console.
+- All HTTP API error responses SHOULD be surfaced via the library unless there is a specific reason that a HTTP API error MUST NOT be surfaced for security reasons.
+
+## Tests and code coverage
+
+The library MUST include a standardised set of unit tests, covering its critical functionality. The exact implementation SHOULD be dictated based on development language best practices.
+
+### Triggering events
+
+- `testPostBatchEventsToChannelFailsForMultichannelEvents`
```

Comment:
Should there be more details for what each test is and expects?

Reply:
I _could_ add some additional details I suppose, however I didn't want to be overly prescriptive and I thought the test names (which are taken from the Channels Swift HTTP SDK) described the high-level behaviour of each test case. What do you think?

Reply:
Perhaps a note along the lines of "the tests are self descriptive, but examples can be found in pusher-http-swift".

Reply:
This is a good compromise - I'll push a change now and then get this merged in :)